### ISSUE TITLE
Constant bitrate for live transmission support

### ISSUE THREAD

**Issue by cpacifying** [Haivision/srt#452]
Hello.
First, I am sorry if this subject was already discussed before. If it was, I was unable to find corresponding threads.

SRT transmission has configurable maximum bitrate, but actual bitrate may vary greatly between `original stream bitrate` and `original stream bitrate * (1 + bandwitdth overhead)`. I have looked into C API manual to find means to stabilise bitrate at its maximum permissible value, but did not succeed.

So, my question is — is there a way to configure SRT channel to occupy fixed bitrate for all duration of transmission? If not, is such feature being planned for implementation?

I have found [this issue](https://github.com/Haivision/srt/issues/292) and related pull requests but little else, and I am not even entirely sure this is related to my question.

**Comment by jeandube** [comment]
I don't think CBR can be implemented at the transport level if the live input stream is not CBR. By default the output rate of the live-mode SRT sender is tailored on the input rate (the output of a streamer). The input rate (sendmsg API) is measured internally and the output pace is adjusted accordingly including the configured overhead (default 25%) added to permit the injection of retransmitted packets. It has been established that the measured rate lags behind the real input rate and this is a problem when video motion emerges from a black screen or still image. The nominal bit rate of the streamer can be set in the SRT session with the SRTO_INPUTBW option (default=0: which means internally measured). This option can be set while the session is active. WARNING: check units from memory it is Bytes/Sec.

While investigating a problem with SRT sender output packet pace lately it appears that this mechanism is not behaving as expected (the bandwidth ceiling is not respected), probably due to the fact that the packet stream of a live input is not as dense as the one of a file transfer, A condition for which the UDT original algorithm has not envision. The playback is not bad on the receiver side but the bandwidth used is not under control.
To summarize: the SRT stream bitrate fluctuate with the input bitrate and the level of packet loss causing retransmission. The INPUTBW and OVERHEAD settings attempts to put a ceiling on the bandwidth used. Even if the output pace mechnism was working well, fluctuation is expected for lost packets retransmission.

**Comment by ethouris** [comment]
What SRT currently does is to keep the bandwidth under ceiling. The user must grant, however, that the input bitrate is below that bandwidth so there's some extra space for prospective retransmission (also as overhead bandwidth), otherwise the rubber-tape-stretching problem will happen.

For me "fixed bitrate" sounds like you'd like to make the bitrate increasing at the moment when the input bitrate decreases. Not sure if I correctly understand your meaning. That could be a problem even if you ensure fixed bitrate on the input stream because of - at least prospective - retransmissions.

**Comment by alexpokotilo** [comment]
A little ads if you don't mind :)
As SRT mostly use mpegts as transport layer the answer is yes you can archive constant bitrate for any streams if you add necessary paddings to your mpegts transport stream in case if source stream' bitrate is below desired value.
In Nimble streamer we implemented the same method we already had for pure udp mpegts delivery.
Please take a look at https://blog.wmspanel.com/2017/07/setup-srt-secure-reliable-transport-nimble-streamer.html for "Mux rate" parameter description.
P.S
As SRT now supported in ffmpeg and ffmpeg supports mux rate via -muxrate parameter I'm 99% sure you can use ffmpeg+SRT+muxrate param to have CBR in transport layer. ffmpeg has good CBR support for UDP+mpegts so I'm almost certain SRT is supported as well.
And I'm sure that you can get an idea from libavformat/mpegtsenc.c file in ffmpeg source package if you want to implement this method in your project without ffmpeg.

**Comment by cpacifying** [comment]
@alexpokotilo, thank you! That might be just what I am looking for.
